server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java

```diff
@@ -2783,84 +2786,140 @@ public ServiceOffering updateServiceOffering(final UpdateServiceOfferingCmd cmd)
-                _serviceOfferingDetailsDao.remove(sc);
-                for (Long domainId : filteredDomainIds) {
-                    detailsVO.add(new ServiceOfferingDetailsVO(id, ApiConstants.DOMAIN_ID, String.valueOf(domainId), false));
-                }
-            }
-            if(!filteredZoneIds.equals(existingZoneIds)) {
-                sc.setParameters("detailName", ApiConstants.ZONE_ID);
-                _serviceOfferingDetailsDao.remove(sc);
-                for (Long zoneId : filteredZoneIds) {
-                    detailsVO.add(new ServiceOfferingDetailsVO(id, ApiConstants.ZONE_ID, String.valueOf(zoneId), false));
-                }
+
+        if (hostTag != null) {
+            if (hostTag.trim().isEmpty()) {
+                fields.add("removing host tag");
+                offering.setHostTag(null);
+            } else {
+                fields.add("host tag: " + hostTag);
+                offering.setHostTag(hostTag);
             }
         }
-        if (!detailsVO.isEmpty()) {
-            for (ServiceOfferingDetailsVO detailVO : detailsVO) {
-                _serviceOfferingDetailsDao.persist(detailVO);
-            }
+
+        s_logger.debug(String.format("Updating service offering ID [%s] with [%s].", serviceOfferingId, String.join(", ", fields)));
+
+        if (!_serviceOfferingDao.update(serviceOfferingId, offering)) {
+            s_logger.warn(String.format("Can't update service offering ID [%s] with [%s].", serviceOfferingId, String.join(", ", fields)));
+        }
+    }
+
+    /**
+     * Update Service Offering Details for specific domains and zones.
+     */
+    private void updateServiceOfferingDetails(final Long serviceOfferingId, List<Long> existingDomainIds, List<Long> existingZoneIds, List<Long> filteredDomainIds,
+            List<Long> filteredZoneIds) {
+        boolean detailsUpdateNeeded = !filteredDomainIds.equals(existingDomainIds) || !filteredZoneIds.equals(existingZoneIds);
+        if (!detailsUpdateNeeded) {
+            s_logger.debug(String.format("No need to update details for service offering ID [%s] for filtered domain IDs [%s], existing domain IDs [%s], "
+                    + "filtered zone IDs [%s], and existing zone IDs [%s].", serviceOfferingId, filteredDomainIds, existingDomainIds, filteredZoneIds, existingZoneIds));
+            return;
+        }
+        List<ServiceOfferingDetailsVO> detailsVO = new ArrayList<>();
+
+        removeServiceOfferingsForDomains(serviceOfferingId, existingDomainIds, filteredDomainIds, detailsVO);
+        removeServiceOfferingsForZones(serviceOfferingId, existingZoneIds, filteredZoneIds, detailsVO);
+
+        if (detailsVO.isEmpty()) {
+            s_logger.info(String.format("No need to update service offering details for service offering ID [%s], because the details remained the same.", serviceOfferingId));
+            return;
+        }
+
+        for (ServiceOfferingDetailsVO detailVO : detailsVO) {
```

Comment:
here's why I want that plural. so `detailsVO` and `detailsVOs`, It is a more difference than `detailVO` and `detailsVO` .